File: cordova-plugin/ios/SP.m
#import <Cordova/CDVPlugin.h>

#include "json-sqlite-engine-wrapper-pro.h"

@interface SP : CDVPlugin

- (void) openDatabaseHandle:(CDVInvokedUrlCommand *)c;
- (void) bb:(CDVInvokedUrlCommand *)c;

@end

@implementation SP

- (void) openDatabaseHandle:(CDVInvokedUrlCommand *)c
{
  if ([c.arguments count] == 0) {
    [self.commandDelegate
      sendPluginResult :
          [CDVPluginResult resultWithStatus : CDVCommandStatus_OK
                            messageAsArray  : @[[NSNull null]]]
      callbackId : c.callbackId];
    return;
  }

  const jsql_handle_t mydbhandle = jsql_pro_db_open(JSQL_PRO_API_VERSION, [[c.arguments objectAtIndex: 0] UTF8String], JSQL_OPEN_READWRITE | JSQL_OPEN_CREATE | JSQL_OPEN_MEMORY);

  [self.commandDelegate
    sendPluginResult :
        [CDVPluginResult resultWithStatus : CDVCommandStatus_OK
                          messageAsArray  : @[@(mydbhandle)]]
    callbackId : c.callbackId];
}

- (void) jsonExecuteStatement:(CDVInvokedUrlCommand *)c
{
  if ([c.arguments count] < 2) {
    [self.commandDelegate
      sendPluginResult :
          [CDVPluginResult resultWithStatus : CDVCommandStatus_OK
                            messageAsArray  : @[[NSNull null]]]
      callbackId : c.callbackId];
    return;
  }

  const char * rj = jsql_pro_execute([[c.arguments objectAtIndex: 0] longLongValue], [[c.arguments objectAtIndex: 1] UTF8String], 0);

  [self.commandDelegate
    sendPluginResult :
        [CDVPluginResult resultWithStatus : CDVCommandStatus_OK
                          messageAsArray  : @[@(rj)]]
    callbackId : c.callbackId];

  jsql_pro_internal_cleanup([[c.arguments objectAtIndex: 0] longLongValue]);
}

@end

File: cordova-plugin/android/io/sqlc/SP.java
package io.sqlc;

public class SP extends org.apache.cordova.CordovaPlugin {
  static final String ndkLibraryName = "json-sqlite-engine-procore-wip-ndk";

  static boolean isLibLoaded = false;

  public SP() /* throws a RuntimeException in case of an issue with the lib */ {
    if (!isLibLoaded) {
      System.loadLibrary(ndkLibraryName);
      isLibLoaded = true;
    }
  }

  @Override
  public boolean execute(String a, org.json.JSONArray aa, org.apache.cordova.CallbackContext cc) {
    try {
      switch(a) {
        case "openDatabaseHandle":
          // FUTURE TBD get database name from arguments
          final String name = ":memory:";
          final long openDatabaseHandle = JSQLCOREPRONDKWrapper.jsql_pro_db_open(
              JSQLCOREPRONDKWrapper.JSQL_PRO_API_VERSION,
              name,
              JSQLCOREPRONDKWrapper.JSQL_OPEN_READWRITE | JSQLCOREPRONDKWrapper.JSQL_OPEN_CREATE | JSQLCOREPRONDKWrapper.JSQL_OPEN_MEMORY);
          final org.json.JSONArray openresult = new org.json.JSONArray();
          openresult.put(openDatabaseHandle);
          cc.success(openresult);
          return true;
        case "jsonExecuteStatement":
          if (aa.length() < 2) return false;
          final long jsonDatabaseHandle = aa.getLong(0);
          final String js = aa.getString(1);
          final String jr = JSQLCOREPRONDKWrapper.jsql_pro_execute(jsonDatabaseHandle, js, 0);
          final org.json.JSONArray r = new org.json.JSONArray();
          r.put(jr);
          JSQLCOREPRONDKWrapper.jsql_pro_internal_cleanup(jsonDatabaseHandle);
          cc.success(r);
          return true;
      }
    } catch(Exception e) { return false; }
    return false;
  }
}
